File: Data-Structures/Linked-Lists/0203_remove_element.py
#linked_list 

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: Optional[ListNode]
        :type val: int
        :rtype: Optional[ListNode]
        """
        #if dummy.val = val, remove 
        ans = ListNode(0, head)
        dummy = ans

        while dummy: 
            while dummy.next and dummy.next.val == val:
                dummy.next = dummy.next.next 
            dummy = dummy.next 
            
        return ans.next

File: Data-Structures/Linked-Lists/0203_remove_element.c
// Practice for: 0203_remove_element.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val; 
    struct ListNode *next;
}; 

struct ListNode *removeElements (struct ListNode *head, int val){
    struct ListNode *dummy = (struct ListNode*)malloc(sizeof(struct ListNode)); 
    dummy->next = head;
    struct ListNode *current = dummy;
    
    while (current->next != NULL){
        if (current->next->val == val){
            //must remove from memory
            struct ListNode *trash = current->next;
            current->next = current->next->next;
            free(trash);
        } else {
            current = current->next;
        }
    }
    //must initialise new variable, free dummy variable
    struct ListNode *newHead = dummy->next;
    free(dummy);
    return newHead;

}
